### PR TITLE
TCP Keepalive added as TcpSetting

### PR DESCRIPTION
**Pull request by zmhh**
Pull request for issue: TCP Keep-Alive #234

### CODE REVIEW

File: src/Net/SocketExtensions.cs [hunk 1]
```diff
@@ -22,9 +22,39 @@ namespace Amqp
...
+            ULONG onoff;
+            ULONG keepalivetime;
+            ULONG keepaliveinterval;
+            };
+            */
+
+            try
+            {
+                ulong dummy = 0;
+                byte[] inOptionValues = new byte[Marshal.SizeOf(dummy) * 3];
```
Comment (xinchen10) [inline src/Net/SocketExtensions.cs]: Marshal is not available on netstandard (.Net Core). We can get the byte[] of each field and merge them into one.

File: src/Net/TcpSettings.cs [hunk 2]
```diff
@@ -31,6 +32,17 @@ public class TcpSettings
         int? sendBufferSize;
         int? sendTimeout;
 
+        /// <summary>
+        /// Specifies the KeepAlive option of a TCP socket.
+        /// Tuple.Item1: KeepAliveTime.
+        /// Tuple.Item2: KeepAliveInterval.
+        /// </summary>
+        public Tuple<ulong, ulong> KeepAlive
```
Comment (xinchen10) [inline src/Net/TcpSettings.cs]: Config names should be explicit. It is not easy to figure out what item1 and item2 mean in a tuple. We can define a TcpKeepAlive class similar to the one you have in comments in the method implementation, or maybe just two properties for the time and interval on TcpSettings.

File: src/Net/TcpSettings.cs [hunk 3]
```diff
@@ -87,6 +99,7 @@ public int SendTimeout
 
         internal void Configure(Socket socket)
         {
+            if (this.KeepAlive != null) socket.SetTcpKeepAlive(this.KeepAlive.Item1, this.KeepAlive.Item2);
```
Comment (xinchen10) [inline src/Net/TcpSettings.cs]: Return value of SetTcpKeepAlive is ignored, which results in non-deterministic behavior of the API. The user may think the setting is in effect but it actually failed to apply. We should just let the set method throw so user's code can handle the error.

File: src/Amqp.Net.csproj [hunk 4]
```diff
@@ -72,6 +72,7 @@
     <Compile Include="Listener\Context.cs" />
     <Compile Include="Listener\ListenerLink.cs" />
     <Compile Include="Net\IConnectionFactory.cs" />
+    <Compile Include="Net\TcpKeepAliveSettings.cs" />
```
Comment (xinchen10) [inline src/Amqp.Net.csproj]: Also need to include it in net40 project.

File: src/Net/TcpKeepAliveSettings.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (xinchen10) [inline src/Net/TcpKeepAliveSettings.cs]: using statements not needed. Remove them.

File: src/Net/TcpKeepAliveSettings.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace Amqp
+{
+    public class TcpKeepAliveSettings
```
Comment (xinchen10) [inline src/Net/TcpKeepAliveSettings.cs]: XML comments for public visible types.

**Comment by zmhh** [comment]
Trying to make sure that what is currently passed in "inOptionValues" is correct...

File: src/Net/TcpKeepAliveSettings.cs [hunk 7]
```diff
@@ -0,0 +1,20 @@
+﻿namespace Amqp
+{
+    /// <summary>
+    /// TCP Keep-Alive settings.
+    /// </summary>
+    public class TcpKeepAliveSettings
+    {
+        public ulong KeepAliveTime
```
Comment (xinchen10) [inline src/Net/TcpKeepAliveSettings.cs]: Can you also add some text for this and the other one? Right now warnings are treated as errors so the XML doc missing warning on these two is causing build error.

File: src/Net/SocketExtensions.cs [hunk 8]
```diff
@@ -22,9 +22,32 @@ namespace Amqp
...
+            ULONG keepaliveinterval;
+            };
+            */
+
+            ulong keepAliveTime = settings.KeepAliveTime;
+            ulong keepAliveInterval = settings.KeepAliveInterval;
+
+            int bytesPerULong = 8;
+            byte[] inOptionValues = new byte[bytesPerULong * 3];
+            BitConverter.GetBytes((ulong)1).CopyTo(inOptionValues, 1);
```
Comment (xinchen10) [inline src/Net/SocketExtensions.cs]: Just curious, why copy to destination array at index 1?

Comment (zmhh) [reply src/Net/SocketExtensions.cs]: I was just looking at it wrong.. reverted this in a following commit

**Pull request merged**